# Guided Capstone Step 4. Pre-Processing and Training Data Development

**The Data Science Method**  


1.   Problem Identification 


2.   Data Wrangling 
  
 
3.   Exploratory Data Analysis   

4.   **Pre-processing and Training Data Development**  
 * Create dummy or indicator features for categorical variables
  * Standardize the magnitude of numeric features
  * Split into testing and training datasets
  * Apply scaler to the testing set
5.   Modeling 
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

**<font color='teal'> Start by loading the necessary packages as we did in step 3 and printing out our current working directory just to confirm we are in the correct project directory. </font>**

In [1]:
import sklearn
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import os
import seaborn as sns
from scipy import stats
from scipy.stats import poisson as pmf
from sklearn.cluster import KMeans

**<font color='teal'>  Load the csv file you created in step 3, remember it should be saved inside your data subfolder and print the first five rows.</font>**

In [2]:
df=pd.read_csv('step3_output.csv')

In [3]:
df.head()

Unnamed: 0                           Name       state  summit_elev  \
0           2               Hilltop Ski Area      Alaska         2090   
1           4            Sunrise Park Resort     Arizona        11100   
2           5  Yosemite Ski & Snowboard Area  California         7800   
3           8         Boreal Mountain Resort  California         7700   
4           9                    Dodge Ridge  California         8200   

   vertical_drop  base_elev  trams  fastEight  fastSixes  fastQuads  ...  \
0            294       1796      0        0.0          0          0  ...   
1           1800       9200      0        0.0          0          1  ...   
2            600       7200      0        0.0          0          0  ...   
3            500       7200      0        0.0          0          1  ...   
4           1600       6600      0        0.0          0          0  ...   

   SkiableTerrain_ac  Snow Making_ac  daysOpenLastYear  yearsOpen  \
0               30.0            30.0             150.0       36.0   
1              800.0            80.0             115.0       49.0   
2               88.0            80.0             110.0       84.0   
3              380.0           200.0             150.0       54.0   
4              862.0           200.0             150.0       69.0   

   averageSnowfall  AdultWeekday  AdultWeekend  projectedDaysOpen  \
0             69.0          30.0      34.00000              152.0   
1            250.0          74.0      78.00000              104.0   
2            300.0          47.0      47.00000              107.0   
3            400.0          49.0      64.16681              150.0   
4            350.0          78.0      78.00000              140.0   

   NightSkiing_ac  clusters  
0            30.0         0  
1            80.0         1  
2             0.0         1  
3           200.0         1  
4             0.0         1  

[5 rows x 28 columns]

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `state`. Add the dummies back to the dataframe and remove the original column for `state`. </font>**

Hint: you can see an example of how to execute this in Aiden's article on preprocessing [here](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-pre-processing-and-training-data-development-fd2d75182967). 

In [4]:
dfo=pd.get_dummies(df['state'])
dfo.head()

Alaska  Arizona  California  Colorado  Connecticut  Idaho  Illinois  \
0       1        0           0         0            0      0         0   
1       0        1           0         0            0      0         0   
2       0        0           1         0            0      0         0   
3       0        0           1         0            0      0         0   
4       0        0           1         0            0      0         0   

   Indiana  Iowa  Maine  ...  Rhode Island  South Dakota  Tennessee  Utah  \
0        0     0      0  ...             0             0          0     0   
1        0     0      0  ...             0             0          0     0   
2        0     0      0  ...             0             0          0     0   
3        0     0      0  ...             0             0          0     0   
4        0     0      0  ...             0             0          0     0   

   Vermont  Virginia  Washington  West Virginia  Wisconsin  Wyoming  
0        0         0           0              0          0        0  
1        0         0           0              0          0        0  
2        0         0           0              0          0        0  
3        0         0           0              0          0        0  
4        0         0           0              0          0        0  

[5 rows x 35 columns]

In [5]:
df=pd.concat([df,dfo],axis=1)
df=df.drop(['state'],axis=1)

In [6]:
df.columns

Index(['Unnamed: 0', 'Name', 'summit_elev', 'vertical_drop', 'base_elev',
       'trams', 'fastEight', 'fastSixes', 'fastQuads', 'quad', 'triple',
       'double', 'surface', 'total_chairs', 'Runs', 'TerrainParks',
       'LongestRun_mi', 'SkiableTerrain_ac', 'Snow Making_ac',
       'daysOpenLastYear', 'yearsOpen', 'averageSnowfall', 'AdultWeekday',
       'AdultWeekend', 'projectedDaysOpen', 'NightSkiing_ac', 'clusters',
       'Alaska', 'Arizona', 'California', 'Colorado', 'Connecticut', 'Idaho',
       'Illinois', 'Indiana', 'Iowa', 'Maine', 'Maryland', 'Massachusetts',
       'Michigan', 'Minnesota', 'Missouri', 'Montana', 'Nevada',
       'New Hampshire', 'New Jersey', 'New Mexico', 'New York',
       'North Carolina', 'Ohio', 'Oregon', 'Pennsylvania', 'Rhode Island',
       'South Dakota', 'Tennessee', 'Utah', 'Vermont', 'Virginia',
       'Washington', 'West Virginia', 'Wisconsin', 'Wyoming'],
      dtype='object')

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing standardize the scale of the features of the dataframe except the name of the resort which we done't need in the dataframe for modeling, so it can be droppped here as well. Also, we want to hold out our response variable(s) so we can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as our response for scaling and modeling. Later we will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave them in the development dataframe. </font>**

In [7]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y=df['AdultWeekend']

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 

## Split into training and testing datasets

**<font color='teal'> Using sklearn model selection import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. We will start by using the adult weekend ticket price as our response variable for modeling.</font>**

In [8]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled,y, test_size=0.25, random_state=1)

Here we start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5. Modeling


This is the fifth step in the Data Science Method. In the previous steps you cleaned and prepared the datasets. Now it's time to get into the most exciting part: modeling! In this exercise, you'll build three different models and compare each model's performance. In the end, you'll choose the best model for demonstrating insights to Big Mountain management.



### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with a Training Dataset

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [9]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

Hint: you will have to use the `predict()` method here as it's used in this [DSM article](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-modeling-56b4233cad1b) about modeling. 

In [10]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred = model.predict(X_test)

## Review Model Outcomes — Iterate over additional models as needed

In [ ]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
# You can plug y_test and y_pred into the functions to evaluate the model

In [11]:
explained_variance_score(y_test,y_pred)

-2.816819134922425e+21

In [12]:
mean_absolute_error(y_test,y_pred)

126386205025.42174

**<font color='teal'> Print the intercept value from the linear model. </font>**

Hint: our linear regression model `lm` has an attribute `intercept_` for the intercept

In [13]:
lm.intercept_

-31596551197.88427

**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the resulting `AdultWeekend` value. Also, because we took the time to scale our x values in the training data, we can compare each of the coeeficients for the features to determine the feature importances. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 


Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [14]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
df1 = pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])
df1.sort_values(by='Coefficient',ascending=False)

Coefficient
fastEight          2.280177e+13
trams              2.197783e+13
fastSixes          2.144998e+13
total_chairs       1.497062e+13
double             8.079706e+12
surface            7.774665e+12
triple             6.678470e+12
quad               3.794212e+12
fastQuads          3.089179e+12
New York           1.366476e+12
Michigan           1.165622e+12
Pennsylvania       1.085965e+12
New Hampshire      1.042976e+12
California         1.042976e+12
Wisconsin          1.042976e+12
Minnesota          8.977019e+11
Vermont            8.977019e+11
Idaho              7.821587e+11
New Mexico         7.821587e+11
Connecticut        7.161573e+11
Maine              7.161573e+11
Wyoming            7.161573e+11
Massachusetts      6.424656e+11
Colorado           6.424656e+11
North Carolina     6.424656e+11
Ohio               6.424656e+11
Iowa               5.580450e+11
Illinois           5.580450e+11
West Virginia      5.580450e+11
Oregon             5.580450e+11
Utah               5.580450e+11
Washington         5.580450e+11
Virginia           4.569879e+11
Indiana            4.569879e+11
Missouri           4.569879e+11
Montana            4.569879e+11
Alaska             3.240882e+11
Rhode Island       3.240882e+11
Maryland           3.240882e+11
Tennessee          3.240882e+11
Arizona            3.240882e+11
South Dakota       3.240882e+11
Nevada             3.240882e+11
New Jersey         8.909028e+10
summit_elev        4.529843e+01
base_elev          3.863291e+01
Unnamed: 0         8.977681e+00
AdultWeekday       7.351562e+00
vertical_drop      6.218727e+00
clusters           2.116699e+00
Runs               1.900391e+00
averageSnowfall    1.199219e+00
SkiableTerrain_ac  1.102539e+00
NightSkiing_ac     1.092773e+00
Snow Making_ac     1.051758e+00
projectedDaysOpen  9.370117e-01
daysOpenLastYear   8.398438e-01
TerrainParks       5.820312e-01
yearsOpen          4.580078e-01
LongestRun_mi      1.530762e-01

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that we care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

Hint: Try to construct another model using exactly the steps we followed above. 

#### Model 2

In [15]:
X=df.drop(['Name','AdultWeekend','Alaska', 'Arizona', 'California',
       'Colorado', 'Connecticut', 'Idaho', 'Illinois', 'Indiana', 'Iowa',
       'Maine', 'Maryland', 'Massachusetts', 'Michigan', 'Minnesota',
       'Missouri', 'Montana', 'Nevada', 'New Hampshire', 'New Jersey',
       'New Mexico', 'New York', 'North Carolina', 'Ohio', 'Oregon',
       'Pennsylvania', 'Rhode Island', 'South Dakota', 'Tennessee', 'Utah',
       'Vermont', 'Virginia', 'Washington', 'West Virginia', 'Wisconsin',
       'Wyoming'],axis=1)
y = df.AdultWeekend

In [16]:
X.shape

(172, 25)

In [17]:
# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 

In [18]:
# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled,y, test_size=0.25, random_state=1)

In [19]:
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)


In [20]:
y_pred = model.predict(X_test)


In [21]:
explained_variance_score(y_test,y_pred)

0.7111659116890887

In [22]:
mean_absolute_error(y_test,y_pred)

6.59647849173989

In [23]:
y_pred

array([55.14525123, 51.5574804 , 50.73470569, 44.28349712, 37.99064398,
       56.22537718, 55.00945184, 50.83463381, 30.9518262 , 57.02805021,
       50.0712843 , 80.14268944, 56.6997193 , 61.64286193, 52.6474245 ,
       70.04666599, 49.60677249, 49.82026331, 89.84773821, 50.9527671 ,
       61.00826119, 65.67886908, 47.62774479, 92.03656998, 65.74430234,
       47.96317127, 43.51820143, 38.22532541, 58.8810183 , 54.97979805,
       77.32478768, 73.44184473, 58.77235436, 66.27964706, 65.42955716,
       54.35129863, 74.34077142, 62.29396288, 41.62263211, 49.8378346 ,
       56.08500578, 54.36172884, 63.26707757])

In [24]:
df2 = pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])
df2.sort_values(by='Coefficient',ascending=False)

Coefficient
base_elev          1.279800e+01
summit_elev        1.194640e+01
AdultWeekday       8.957812e+00
clusters           3.449601e+00
daysOpenLastYear   2.359202e+00
projectedDaysOpen  2.030769e+00
averageSnowfall    1.611199e+00
Snow Making_ac     1.439888e+00
quad               1.306382e+00
Runs               1.267756e+00
TerrainParks       1.238383e+00
vertical_drop      9.750988e-01
triple             9.533158e-01
NightSkiing_ac     9.319982e-01
total_chairs       8.456099e-01
SkiableTerrain_ac  7.230765e-01
double             6.623160e-01
surface            5.457131e-01
yearsOpen          2.522711e-01
LongestRun_mi      2.122933e-01
fastQuads          7.359212e-02
Unnamed: 0         4.218960e-03
fastEight          1.998401e-14
fastSixes          1.787459e-14
trams              1.509903e-14

**<font color='teal'> When reviewing our new model coefficients, we see `summit_elev` is now in the number two spot. This is also difficult to change from a management prespective and highly correlated with `base_elev` and `vertical_drop`.  This time, rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

#### Model 3

In [25]:
X = df.drop(['Name','AdultWeekend','Alaska', 'Arizona', 'California', 'Colorado',
       'Connecticut', 'Idaho', 'Illinois', 'Indiana', 'Iowa', 'Maine',
       'Maryland', 'Massachusetts', 'Michigan', 'Minnesota', 'Missouri',
       'Montana', 'Nevada', 'New Hampshire', 'New Jersey', 'New Mexico',
       'New York', 'North Carolina', 'Ohio', 'Oregon', 'Pennsylvania',
       'Rhode Island', 'South Dakota', 'Tennessee', 'Utah', 'Vermont',
       'Virginia', 'Washington', 'West Virginia', 'Wisconsin', 'Wyoming','summit_elev'], axis=1)

y = df.AdultWeekend

In [26]:
# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 

In [27]:
# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled,y, test_size=0.25, random_state=1)
#model 3 
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

In [28]:
y_pred = model.predict(X_test)

In [29]:
explained_variance_score(y_test,y_pred)

0.7168960734960592

In [30]:
mean_absolute_error(y_test,y_pred)

6.5095652467905945

In [31]:
df3 = pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])
df3.sort_values(by='Coefficient',ascending=False)

Coefficient
AdultWeekday       9.079836e+00
clusters           3.433655e+00
daysOpenLastYear   2.381985e+00
base_elev          1.938673e+00
projectedDaysOpen  1.900117e+00
averageSnowfall    1.723620e+00
quad               1.434700e+00
Snow Making_ac     1.312316e+00
vertical_drop      1.225406e+00
TerrainParks       1.210079e+00
Runs               1.200925e+00
triple             9.302911e-01
NightSkiing_ac     9.181306e-01
total_chairs       8.098623e-01
SkiableTerrain_ac  7.766941e-01
double             6.434429e-01
surface            6.267216e-01
yearsOpen          2.306039e-01
LongestRun_mi      1.680548e-01
Unnamed: 0         5.493826e-02
fastQuads          4.576695e-02
trams              2.220446e-15
fastEight          1.998401e-15
fastSixes          2.220446e-16

## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell — you will discuss this selection more in the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. | -2.816819134922425e+21 |126386205025.42174 |-|
| Model 2. |0.7111659116890887| 6.59647849173989|'state'|
| Model 3. |0.7168960734960592 | 6.5095652467905945|'state','summit_elev'|

Model Selection:

In [ ]:
#Model 2 